File: 2024/javascript/07-theGrinchsAttack.js
// 07. The Grinch's Attack

/** @param {string} packages with parentheses
 *  @returns {string} Fixed and sorted packages
 */
function fixPackages(packages) {
  const regex = /\(\w+\)/;

  while (regex.test(packages)) {
    packages = packages.replace(regex, (match) =>
      [...match].slice(1, -1).reverse().join("")
    );
  }
  return packages;
}

console.log(fixPackages("a(cb)de"));
// ➞ "abcde"
// We reverse "cb" inside the parentheses

console.log(fixPackages("a(bc(def)g)h"));
// ➞ "agdefcbh"
// 1st we reverse "def" → "fed", then we reverse "bcfedg" → "gdefcb"

console.log(fixPackages("abc(def(gh)i)jk"));
// ➞ "abcighfedjk"
// 1st we reverse "gh" → "hg", then "defhgi" → "ighfed"

console.log(fixPackages("a(b(c))e"));
// ➞ "acbe"
// 1st we reverse "c" → "c", then "bc" → "cb"

File: 2024/python/07_the_grinchs_attack.py
"""07. The Grinch's Attack"""

import re


def fix_packages(packages: str) -> str:
    """Fix Packages.

    Args:
        packages (str): Packages with parentheses

    Returns:
        str: Fixed and sorted packages
    """
    regex = r"\(\w+\)"

    while match := re.search(regex, packages):
        packages = re.sub(regex, match.group()[1:-1][::-1], packages)

    return packages


print(fix_packages("a(cb)de"))
# ➞ "abcde"
# We reverse "cb" inside the parentheses

print(fix_packages("a(bc(def)g)h"))
# ➞ "agdefcbh"
# 1st we reverse "def" → "fed", then we reverse "bcfedg" → "gdefcb"

print(fix_packages("abc(def(gh)i)jk"))
# ➞ "abcighfedjk"
# 1st we reverse "gh" → "hg", then "defhgi" → "ighfed"

print(fix_packages("a(bc(def)g)h"))
# ➞ "acbe"
# 1st we reverse "c" → "c", then "bc" → "cb"
